[android]: AssertionError: Must be called only after 'onEnteredIntoFirstRun' method!

```
Fatal Exception: java.lang.AssertionError: Must be called only after 'onEnteredIntoFirstRun' method!
       at com.mapswithme.maps.location.LocationHelper.onExitFromFirstRun(LocationHelper.java:600)
       at com.mapswithme.maps.MapFragment$2.run(MapFragment.java:204)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8512)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

https://console.firebase.google.com/u/0/project/omapsapp/crashlytics/app/android:app.omaps.beta/issues/9adc72da7185b2ec6f0cbc6fc1e8b09d

--- a/android/src/com/mapswithme/maps/DownloadResourcesLegacyActivity.java
+++ b/android/src/com/mapswithme/maps/DownloadResourcesLegacyActivity.java
@@ -28,6 +28,7 @@ import com.mapswithme.maps.intent.MapTask;
 import com.mapswithme.maps.location.LocationHelper;
 import com.mapswithme.maps.location.LocationListener;
 import com.mapswithme.util.ConnectionState;
+import com.mapswithme.util.Counters;
 import com.mapswithme.util.StringUtils;
 import com.mapswithme.util.UiUtils;
 import com.mapswithme.util.Utils;
@@ -446,8 +447,8 @@ public class DownloadResourcesLegacyActivity extends BaseMwmFragmentActivity imp
     if (intent == null)
       return null;
 
-    MwmApplication application = MwmApplication.from(this);
-    intent.putExtra(Factory.EXTRA_IS_FIRST_LAUNCH, application.isFirstLaunch());
+    boolean isFirstLaunch = Counters.isFirstLaunch(this);
+    intent.putExtra(Factory.EXTRA_IS_FIRST_LAUNCH, isFirstLaunch);
     if (intent.getData() == null)
     {
       /* Sic: the deep link library has been removed */

--- a/android/src/com/mapswithme/maps/MapFragment.java
+++ b/android/src/com/mapswithme/maps/MapFragment.java
@@ -20,6 +20,8 @@ import androidx.appcompat.app.AlertDialog;
 import com.mapswithme.maps.base.BaseMwmFragment;
 import com.mapswithme.maps.location.LocationHelper;
 import com.mapswithme.util.Config;
+import com.mapswithme.util.Counters;
+import com.mapswithme.util.PermissionsUtils;
 import com.mapswithme.util.UiUtils;
 import com.mapswithme.util.concurrency.UiThread;
 import com.mapswithme.util.log.Logger;
@@ -186,7 +188,7 @@ public class MapFragment extends BaseMwmFragment
     requireActivity().getWindowManager().getDefaultDisplay().getMetrics(metrics);
     final float exactDensityDpi = metrics.densityDpi;
 
-    final boolean firstStart = MwmApplication.from(requireActivity()).isFirstLaunch();
+    final boolean firstStart = LocationHelper.INSTANCE.isInFirstRun();
     if (!nativeCreateEngine(surface, (int) exactDensityDpi, firstStart, mLaunchByDeepLink,
                             BuildConfig.VERSION_CODE))
     {

--- a/android/src/com/mapswithme/maps/MwmActivity.java
+++ b/android/src/com/mapswithme/maps/MwmActivity.java
@@ -99,6 +99,7 @@ import com.mapswithme.maps.widget.placepage.PlacePageController;
 import com.mapswithme.maps.widget.placepage.PlacePageData;
 import com.mapswithme.maps.widget.placepage.PlacePageFactory;
 import com.mapswithme.maps.widget.placepage.RoutingModeListener;
+import com.mapswithme.util.Counters;
 import com.mapswithme.util.InputUtils;
 import com.mapswithme.util.PermissionsUtils;
 import com.mapswithme.util.SharingUtils;
@@ -455,10 +456,11 @@ public class MwmActivity extends BaseMwmFragmentActivity
     initViews(isLaunchByDeepLink);
 
     boolean isConsumed = savedInstanceState == null && processIntent(getIntent());
+    boolean isFirstLaunch = Counters.isFirstLaunch(this);
     // If the map activity is launched by any incoming intent (deeplink, update maps event, etc)
     // or it's the first launch (onboarding) we haven't to try restoring the route,
     // showing the tips, etc.
-    if (isConsumed || MwmApplication.from(this).isFirstLaunch())
+    if (isConsumed || isFirstLaunch)
       return;
 
     if (savedInstanceState == null && RoutingController.get().hasSavedRoute())

--- a/android/src/com/mapswithme/maps/MwmApplication.java
+++ b/android/src/com/mapswithme/maps/MwmApplication.java
@@ -40,6 +40,7 @@ import com.mapswithme.util.Utils;
 import com.mapswithme.util.log.Logger;
 import com.mapswithme.util.log.LoggerFactory;
 
+import java.io.IOException;
 import java.util.List;
 
 public class MwmApplication extends Application implements AppBackgroundTracker.OnTransitionListener
@@ -123,7 +124,17 @@ public class MwmApplication extends Application implements AppBackgroundTracker.
     getLogger().d(TAG, "Application is created");
     // Set configuration directory as early as possible.
     // Other methods may explicitly use Config, which requires settingsDir to be set.
-    setSettingsDir();
+    final String settingsPath = StorageUtils.getSettingsPath(this);
+    getLogger().d(TAG, "Settings path = " + settingsPath);
+    try
+    {
+      StorageUtils.createDirectory(settingsPath);
+      nativeSetSettingsDir(settingsPath);
+    } catch (IOException e)
+    {
+      // We have nothing to do here.
+      throw new AssertionError("Can't create settingsDir " + settingsPath);
+    }
     mMainLoopHandler = new Handler(getMainLooper());
     ConnectionState.INSTANCE.initialize(this);
     CrashlyticsUtils.INSTANCE.initialize(this);
@@ -145,36 +156,19 @@ public class MwmApplication extends Application implements AppBackgroundTracker.
   }
 
   /**
-   * Initialize configuration directory.
-   */
-  public void setSettingsDir()
-  {
-    final String settingsPath = StorageUtils.getSettingsPath(this);
-    if (!StorageUtils.createDirectory(this, settingsPath))
-    {
-      throw new AssertionError("Can't create settingsDir " + settingsPath);
-    }
-    getLogger().d(TAG, "Settings path = " + settingsPath);
-    nativeSetSettingsDir(settingsPath);
-  }
-
-  /**
-   * Initialize native core of application: platform and framework. Caller must handle returned value
-   * and do nothing with native code if initialization is failed.
+   * Initialize native core of application: platform and framework.
    *
    * @return boolean - indicator whether native initialization is successful or not.
+   * @throws IOException - if failed to create directories. Caller must handle
+   * the exception and do nothing with native code if initialization is failed.
    */
-  public boolean initCore()
+  public void ensureCoreInitialized() throws IOException
   {
     initNativePlatform();
-    if (!mPlatformInitialized)
-      return false;
-
     initNativeFramework();
-    return mFrameworkInitialized;
   }
 
-  private void initNativePlatform()
+  private void initNativePlatform() throws IOException
   {
     if (mPlatformInitialized)
       return;
@@ -193,8 +187,7 @@ public class MwmApplication extends Application implements AppBackgroundTracker.
     // If platform directories are not created it means that native part of app will not be able
     // to work at all. So, we just ignore native part initialization in this case, e.g. when the
     // external storage is damaged or not available (read-only).
-    if (!createPlatformDirectories(writablePath, privatePath, tempPath))
-      return;
+    createPlatformDirectories(writablePath, privatePath, tempPath);
 
     nativeInitPlatform(apkPath,
                        writablePath,
@@ -207,18 +200,18 @@ public class MwmApplication extends Application implements AppBackgroundTracker.
 
     Editor.init(this);
     mPlatformInitialized = true;
+    log.i(TAG, "Platform initialized");
   }
 
-  private boolean createPlatformDirectories(@NonNull String writablePath,
+  private void createPlatformDirectories(@NonNull String writablePath,
                                             @NonNull String privatePath,
-                                            @NonNull String tempPath)
+                                            @NonNull String tempPath) throws IOException
   {
-    if (SharedPropertiesUtils.shouldEmulateBadExternalStorage(this))
-      return false;
+    SharedPropertiesUtils.emulateBadExternalStorage(this);
 
-    return StorageUtils.createDirectory(this, writablePath) &&
-           StorageUtils.createDirectory(this, privatePath) &&
-           StorageUtils.createDirectory(this, tempPath);
+    StorageUtils.createDirectory(writablePath);
+    StorageUtils.createDirectory(privatePath);
+    StorageUtils.createDirectory(tempPath);
   }
 
   private void initNativeFramework()
@@ -243,6 +236,8 @@ public class MwmApplication extends Application implements AppBackgroundTracker.
     IsolinesManager.from(this).initialize(null);
     TrackRecorder.INSTANCE.initialize(this);
     mBackgroundTracker.addListener(this);
+
+    getLogger().i(TAG, "Framework initialized");
     mFrameworkInitialized = true;
   }
 
@@ -312,11 +307,6 @@ public class MwmApplication extends Application implements AppBackgroundTracker.
   public Logger getLogger()
   {
     return mLogger;
-  }
-
-  public void setFirstLaunch(boolean isFirstLaunch)
-  {
-    mFirstLaunch = isFirstLaunch;
   }
 
   public boolean isFirstLaunch()

--- a/android/src/com/mapswithme/maps/MwmBroadcastReceiver.java
+++ b/android/src/com/mapswithme/maps/MwmBroadcastReceiver.java
@@ -11,6 +11,8 @@ import com.mapswithme.util.CrashlyticsUtils;
 import com.mapswithme.util.log.Logger;
 import com.mapswithme.util.log.LoggerFactory;
 
+import java.io.IOException;
+
 public abstract class MwmBroadcastReceiver extends BroadcastReceiver
 {
   private static final Logger LOGGER = LoggerFactory.INSTANCE.getLogger(LoggerFactory.Type.MISC);
@@ -30,13 +32,15 @@ public abstract class MwmBroadcastReceiver extends BroadcastReceiver
     String msg = "onReceive: " + intent;
     LOGGER.i(getTag(), msg);
     CrashlyticsUtils.INSTANCE.log(Log.INFO, getTag(), msg);
-    if (!app.arePlatformAndCoreInitialized() && !app.initCore())
+    try
     {
-      LOGGER.w(getTag(), "Application is not initialized, ignoring " + intent);
+      app.ensureCoreInitialized();
+    } catch (IOException e)
+    {
+      LOGGER.e(getTag(), "Failed to initialize application, ignoring " + intent);
       return;
     }
+
     onReceiveInitialized(context, intent);
   }
-
-
 }

--- a/android/src/com/mapswithme/maps/MwmJobIntentService.java
+++ b/android/src/com/mapswithme/maps/MwmJobIntentService.java
@@ -10,6 +10,8 @@ import com.mapswithme.util.CrashlyticsUtils;
 import com.mapswithme.util.log.Logger;
 import com.mapswithme.util.log.LoggerFactory;
 
+import java.io.IOException;
+
 public abstract class MwmJobIntentService extends JobIntentService
 {
   private static final Logger LOGGER = LoggerFactory.INSTANCE.getLogger(LoggerFactory.Type.MISC);
@@ -29,11 +31,15 @@ public abstract class MwmJobIntentService extends JobIntentService
     String msg = "onHandleWork: " + intent;
     LOGGER.i(getTag(), msg);
     CrashlyticsUtils.INSTANCE.log(Log.INFO, getTag(), msg);
-    if (!app.arePlatformAndCoreInitialized() && !app.initCore())
+    try
     {
-      LOGGER.w(getTag(), "Application is not initialized, ignoring " + intent);
+      app.ensureCoreInitialized();
+    } catch (IOException e)
+    {
+      LOGGER.e(getTag(), "Failed to initialize application, ignoring " + intent);
       return;
     }
+
     onHandleWorkInitialized(intent);
   }
 }

--- a/android/src/com/mapswithme/maps/SplashActivity.java
+++ b/android/src/com/mapswithme/maps/SplashActivity.java
@@ -5,50 +5,32 @@ import android.content.Context;
 import android.content.DialogInterface;
 import android.content.Intent;
 import android.os.Bundle;
-import android.view.View;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
+import androidx.annotation.StringRes;
 import androidx.appcompat.app.AlertDialog;
 import androidx.appcompat.app.AppCompatActivity;
-import androidx.fragment.app.FragmentManager;
 
 import com.mapswithme.maps.base.BaseActivity;
 import com.mapswithme.maps.base.BaseActivityDelegate;
 import com.mapswithme.maps.location.LocationHelper;
-import com.mapswithme.util.Config;
 import com.mapswithme.util.Counters;
 import com.mapswithme.util.PermissionsUtils;
 import com.mapswithme.util.ThemeUtils;
 import com.mapswithme.util.UiUtils;
 import com.mapswithme.util.concurrency.UiThread;
-import com.mapswithme.util.log.Logger;
-import com.mapswithme.util.log.LoggerFactory;
+
+import java.io.IOException;
 
 public class SplashActivity extends AppCompatActivity implements BaseActivity
 {
-  private static final Logger LOGGER = LoggerFactory.INSTANCE.getLogger(LoggerFactory.Type.MISC);
-  private static final String TAG = SplashActivity.class.getSimpleName();
   private static final String EXTRA_ACTIVITY_TO_START = "extra_activity_to_start";
   public static final String EXTRA_INITIAL_INTENT = "extra_initial_intent";
   private static final int REQUEST_PERMISSIONS = 1;
   private static final long DELAY = 100;
 
-  private View mIvLogo;
-  private View mAppName;
-
-  private boolean mPermissionsGranted;
-  private boolean mCanceled;
-
-  @NonNull
-  private final Runnable mPermissionsDelayedTask = new Runnable()
-  {
-    @Override
-    public void run()
-    {
-      PermissionsUtils.requestLocationPermission(SplashActivity.this, REQUEST_PERMISSIONS);
-    }
-  };
+  private boolean mCanceled = false;
 
   @NonNull
   private final Runnable mInitCoreDelayedTask = new Runnable()
@@ -56,31 +38,7 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
     @Override
     public void run()
     {
-      MwmApplication app = (MwmApplication) getApplication();
-      if (app.arePlatformAndCoreInitialized())
-      {
-        UiThread.runLater(mFinalDelayedTask);
-        return;
-      }
-
       init();
-
-      LOGGER.i(TAG, "Core initialized: " + app.arePlatformAndCoreInitialized());
-
-//    Run delayed task because resumeDialogs() must see the actual value of mCanceled flag,
-//    since onPause() callback can be blocked because of UI thread is busy with framework
-//    initialization.
-      UiThread.runLater(mFinalDelayedTask);
-    }
-  };
-
-  @NonNull
-  private final Runnable mFinalDelayedTask = new Runnable()
-  {
-    @Override
-    public void run()
-    {
-      resumeDialogs();
     }
   };
 
@@ -104,11 +62,10 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
   {
     super.onCreate(savedInstanceState);
     mBaseDelegate.onCreate();
-    UiThread.cancelDelayedTasks(mPermissionsDelayedTask);
     UiThread.cancelDelayedTasks(mInitCoreDelayedTask);
-    UiThread.cancelDelayedTasks(mFinalDelayedTask);
     Counters.initCounters(this);
-    initView();
+    UiUtils.setupStatusBar(this);
+    setContentView(R.layout.activity_splash);
   }
 
   @Override
@@ -130,37 +87,14 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
   {
     super.onResume();
     mBaseDelegate.onResume();
-    mCanceled = false;
-
-    Context context = getApplicationContext();
-    if (Counters.isMigrationNeeded(context))
+    if (mCanceled)
+      return;
+    if (!PermissionsUtils.isLocationGranted(this))
     {
-      Config.migrateCountersToSharedPrefs(context);
-      Counters.setMigrationExecuted(context);
-    }
-    
-    final boolean isFirstLaunch = isFirstLaunch();
-    if (isFirstLaunch)
-      MwmApplication.from(this).setFirstLaunch(true);
-
-    if (processPermissionGranting())
-      runInitCoreTask();
-  }
-
-  private boolean processPermissionGranting()
-  {
-    mPermissionsGranted = PermissionsUtils.isLocationGranted(this);
-    if (!mPermissionsGranted)
-    {
-      UiThread.runLater(mPermissionsDelayedTask, DELAY);
-      return false;
+      PermissionsUtils.requestLocationPermission(SplashActivity.this, REQUEST_PERMISSIONS);
+      return;
     }
 
-    return true;
-  }
-
-  private void runInitCoreTask()
-  {
     UiThread.runLater(mInitCoreDelayedTask, DELAY);
   }
 
@@ -169,10 +103,7 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
   {
     super.onPause();
     mBaseDelegate.onPause();
-    mCanceled = true;
-    UiThread.cancelDelayedTasks(mPermissionsDelayedTask);
     UiThread.cancelDelayedTasks(mInitCoreDelayedTask);
-    UiThread.cancelDelayedTasks(mFinalDelayedTask);
   }
 
   @Override
@@ -189,27 +120,13 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
     mBaseDelegate.onDestroy();
   }
 
-  private void resumeDialogs()
+  private void showFatalErrorDialog(@StringRes int titleId, @StringRes int messageId)
   {
-    if (mCanceled)
-      return;
-
-    MwmApplication app = (MwmApplication) getApplication();
-    if (!app.arePlatformAndCoreInitialized())
-    {
-      showExternalStorageErrorDialog();
-      return;
-    }
-
-    processNavigation();
-  }
-
-  private void showExternalStorageErrorDialog()
-  {
+    mCanceled = true;
     AlertDialog dialog = new AlertDialog.Builder(this)
-        .setTitle(R.string.dialog_error_storage_title)
-        .setMessage(R.string.dialog_error_storage_message)
-        .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener()
+        .setTitle(titleId)
+        .setMessage(messageId)
+        .setNegativeButton(R.string.ok, new DialogInterface.OnClickListener()
         {
           @Override
           public void onClick(DialogInterface dialog, int which)
@@ -226,32 +143,37 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
   public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                          @NonNull int[] grantResults)
   {
+    if (requestCode != REQUEST_PERMISSIONS)
+      throw new AssertionError("Unexpected requestCode");
     super.onRequestPermissionsResult(requestCode, permissions, grantResults);
-    if (grantResults.length == 0)
+    if (!PermissionsUtils.isLocationGranted(this))
+    {
+      showFatalErrorDialog(R.string.enable_location_services, R.string.location_is_disabled_long_text);
       return;
-
-    mPermissionsGranted = PermissionsUtils.computePermissionsResult(permissions, grantResults)
-                                          .isLocationGranted();
-  }
-
-  private void initView()
-  {
-    UiUtils.setupStatusBar(this);
-    setContentView(R.layout.activity_splash);
-    mIvLogo = findViewById(R.id.iv__logo);
-    mAppName = findViewById(R.id.tv__app_name);
+    }
+    // No-op here - onResume() calls init();
   }
 
   private void init()
   {
     MwmApplication app = MwmApplication.from(this);
-    boolean success = app.initCore();
-    if (!success || !app.isFirstLaunch())
+    try
+    {
+      app.ensureCoreInitialized();
+    } catch (IOException e)
+    {
+      showFatalErrorDialog(R.string.dialog_error_storage_title, R.string.dialog_error_storage_message);
       return;
+    }
 
-    LocationHelper.INSTANCE.onEnteredIntoFirstRun();
-    if (!LocationHelper.INSTANCE.isActive())
-      LocationHelper.INSTANCE.start();
+    if (Counters.isFirstLaunch(this))
+    {
+      LocationHelper.INSTANCE.onEnteredIntoFirstRun();
+      if (!LocationHelper.INSTANCE.isActive())
+        LocationHelper.INSTANCE.start();
+    }
+
+    processNavigation();
   }
 
   @SuppressWarnings("unchecked")
@@ -272,6 +194,7 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
                            input;
       result.putExtra(EXTRA_INITIAL_INTENT, initialIntent);
     }
+    Counters.setFirstStartDialogSeen(this);
     startActivity(result);
     finish();
   }
@@ -296,15 +219,5 @@ public class SplashActivity extends AppCompatActivity implements BaseActivity
     throw new IllegalArgumentException("Attempt to apply unsupported theme: " + theme);
   }
 
-  boolean isFirstLaunch()
-  {
-    if (Counters.getFirstInstallVersion(getApplicationContext()) < BuildConfig.VERSION_CODE)
-      return false;
 
-    FragmentManager fm = getSupportFragmentManager();
-    if (fm.isDestroyed())
-      return false;
-
-    return !Counters.isFirstStartDialogSeen(this);
-  }
 }

--- a/android/src/com/mapswithme/maps/background/UpgradeReceiver.java
+++ b/android/src/com/mapswithme/maps/background/UpgradeReceiver.java
@@ -1,28 +1,18 @@
 package com.mapswithme.maps.background;
 
-import static com.mapswithme.maps.MwmApplication.backgroundTracker;
-
-import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.Intent;
-import android.util.Log;
+
+import androidx.annotation.NonNull;
 
 import com.mapswithme.maps.MwmApplication;
-import com.mapswithme.util.CrashlyticsUtils;
-import com.mapswithme.util.log.Logger;
-import com.mapswithme.util.log.LoggerFactory;
+import com.mapswithme.maps.MwmBroadcastReceiver;
 
-public class UpgradeReceiver extends BroadcastReceiver
+public class UpgradeReceiver extends MwmBroadcastReceiver
 {
-  private static final Logger LOGGER = LoggerFactory.INSTANCE.getLogger(LoggerFactory.Type.MISC);
-  private static final String TAG = UpgradeReceiver.class.getSimpleName();
   @Override
-  public void onReceive(Context context, Intent intent)
+  protected void onReceiveInitialized(@NonNull Context context, @NonNull Intent intent)
   {
-    String msg = "onReceive: " + intent + " app in background = "
-                 + !backgroundTracker(context).isForeground();
-    LOGGER.i(TAG, msg);
-    CrashlyticsUtils.INSTANCE.log(Log.INFO, TAG, msg);
     MwmApplication.onUpgrade(context);
   }
 }

--- a/android/src/com/mapswithme/maps/location/LocationHelper.java
+++ b/android/src/com/mapswithme/maps/location/LocationHelper.java
@@ -586,6 +586,12 @@ public enum LocationHelper implements Initializable<Context>
   }
 
   @UiThread
+  public boolean isInFirstRun()
+  {
+    return mInFirstRun;
+  }
+
+  @UiThread
   public void onEnteredIntoFirstRun()
   {
     mLogger.i(TAG, "onEnteredIntoFirstRun");

--- a/android/src/com/mapswithme/util/Counters.java
+++ b/android/src/com/mapswithme/util/Counters.java
@@ -6,6 +6,7 @@ import android.text.format.DateUtils;
 
 import androidx.annotation.NonNull;
 import androidx.fragment.app.DialogFragment;
+import androidx.fragment.app.FragmentManager;
 import androidx.preference.PreferenceManager;
 
 import com.mapswithme.maps.BuildConfig;
@@ -24,8 +25,6 @@ public final class Counters
   static final String KEY_LIKES_LAST_RATED_SESSION = "LastRatedSession";
 
   private static final String KEY_LIKES_RATED_DIALOG = "RatedDialog";
-  private static final String KEY_SHOW_REVIEW_FOR_OLD_USER = "ShowReviewForOldUser";
-  private static final String KEY_MIGRATION_EXECUTED = "MigrationExecuted";
 
   private Counters() {}
 
@@ -40,26 +39,17 @@ public final class Counters
     return MwmApplication.prefs(context).getInt(KEY_APP_FIRST_INSTALL_VERSION, 0);
   }
 
-  public static boolean isFirstStartDialogSeen(@NonNull Context context)
+  public static boolean isFirstLaunch(@NonNull Context context)
   {
-    return MwmApplication.prefs(context).getBoolean(KEY_MISC_FIRST_START_DIALOG_SEEN, false);
+    return !MwmApplication.prefs(context).getBoolean(KEY_MISC_FIRST_START_DIALOG_SEEN, false);
   }
 
   public static void setFirstStartDialogSeen(@NonNull Context context)
   {
     MwmApplication.prefs(context)
-                  .edit()
-                  .putBoolean(KEY_MISC_FIRST_START_DIALOG_SEEN, true)
-                  .apply();
-  }
-
-
-  public static void setWhatsNewShown(@NonNull Context context)
-  {
-    MwmApplication.prefs(context)
-                  .edit()
-                  .putInt(KEY_MISC_NEWS_LAST_VERSION, BuildConfig.VERSION_CODE)
-                  .apply();
+        .edit()
+        .putBoolean(KEY_MISC_FIRST_START_DIALOG_SEEN, true)
+        .apply();
   }
 
   public static void resetAppSessionCounters(@NonNull Context context)
@@ -169,29 +159,5 @@ public final class Counters
                   .putInt(key, ++value)
                   .apply();
     return value;
-  }
-
-  public static void setShowReviewForOldUser(@NonNull Context context, boolean value)
-  {
-    MwmApplication.prefs(context).edit()
-                  .putBoolean(KEY_SHOW_REVIEW_FOR_OLD_USER, value)
-                  .apply();
-  }
-
-  public static boolean isShowReviewForOldUser(@NonNull Context context)
-  {
-    return MwmApplication.prefs(context).getBoolean(KEY_SHOW_REVIEW_FOR_OLD_USER, false);
-  }
-
-  public static boolean isMigrationNeeded(@NonNull Context context)
-  {
-    return !MwmApplication.prefs(context).getBoolean(KEY_MIGRATION_EXECUTED, false);
-  }
-
-  public static void setMigrationExecuted(@NonNull Context context)
-  {
-    MwmApplication.prefs(context).edit()
-                  .putBoolean(KEY_MIGRATION_EXECUTED, true)
-                  .apply();
   }
 }

--- a/android/src/com/mapswithme/util/PermissionsUtils.java
+++ b/android/src/com/mapswithme/util/PermissionsUtils.java
@@ -57,8 +57,9 @@ public final class PermissionsUtils
     Map<String, Boolean> result = new HashMap<>();
     for (String permission: PERMISSIONS)
     {
-      result.put(permission, Build.VERSION.SDK_INT < Build.VERSION_CODES.M
-                 || appContext.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED);
+      boolean granted = Build.VERSION.SDK_INT < Build.VERSION_CODES.M
+          || appContext.checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED;
+      result.put(permission, granted);
     }
 
     return getPermissionsResult(result);
@@ -72,11 +73,6 @@ public final class PermissionsUtils
                               || (result.containsKey(ACCESS_FINE_LOCATION)
                                   ? result.get(ACCESS_FINE_LOCATION) : false);
     return new PermissionsResult(locationGranted);
-  }
-
-  public static void requestPermissions(@NonNull Activity activity, int code)
-  {
-    ActivityCompat.requestPermissions(activity, PERMISSIONS, code);
   }
 
   public static void requestLocationPermission(@NonNull Activity activity, int code)

--- a/android/src/com/mapswithme/util/SharedPropertiesUtils.java
+++ b/android/src/com/mapswithme/util/SharedPropertiesUtils.java
@@ -13,6 +13,7 @@ import com.mapswithme.maps.MwmApplication;
 import com.mapswithme.maps.R;
 import com.mapswithme.maps.maplayer.Mode;
 
+import java.io.IOException;
 import java.util.Locale;
 
 public final class SharedPropertiesUtils
@@ -63,12 +64,18 @@ public final class SharedPropertiesUtils
     return prefs.getBoolean(PREFS_SHOW_EMULATE_BAD_STORAGE_SETTING, false);
   }
 
-  public static boolean shouldEmulateBadExternalStorage(@NonNull Context context)
+  /**
+   * @param context context
+   * @throws IOException if "Emulate bad storage" is enabled in preferences
+   */
+  public static void emulateBadExternalStorage(@NonNull Context context) throws IOException
   {
     SharedPreferences prefs = PreferenceManager
         .getDefaultSharedPreferences(MwmApplication.from(context));
     String key = MwmApplication.from(context).getString(R.string.pref_emulate_bad_external_storage);
-    return prefs.getBoolean(key, false);
+    if (prefs.getBoolean(key, false)) {
+      throw new IOException("Bad external storage error injection");
+    }
   }
 
   public static boolean shouldShowNewMarkerForLayerMode(@NonNull Context context,

--- a/android/src/com/mapswithme/util/StorageUtils.java
+++ b/android/src/com/mapswithme/util/StorageUtils.java
@@ -17,6 +17,7 @@ import com.mapswithme.util.log.Logger;
 import com.mapswithme.util.log.LoggerFactory;
 
 import java.io.File;
+import java.io.IOException;
 
 public class StorageUtils
 {
@@ -138,17 +139,16 @@ public class StorageUtils
     return addTrailingSeparator(application.getCacheDir().getAbsolutePath());
   }
 
-  public static boolean createDirectory(@NonNull Context context, @NonNull String path)
+  public static void createDirectory(@NonNull String path) throws IOException
   {
     File directory = new File(path);
     if (!directory.exists() && !directory.mkdirs())
     {
-      Throwable error = new IllegalStateException("Can't create directories for: " + path);
+      IOException error = new IOException("Can't create directories for: " + path);
       LOGGER.e(TAG, "Can't create directories for: " + path);
       CrashlyticsUtils.INSTANCE.logException(error);
-      return false;
+      throw error;
     }
-    return true;
   }
 
   static long getFileSize(@NonNull String path)

